### ISSUE TITLE
google-cloud-ruby does not work with Application Default Credentials

### ISSUE THREAD

**Issue by beccasaurus** [googleapis/google-cloud-ruby#951]
See also: [google-cloud-node does not work with Application Default Credentials](https://github.com/GoogleCloudPlatform/google-cloud-node/issues/1653)
## Problem

After running `gcloud beta auth application-default login` I should be able to use google-cloud-ruby to access Google Cloud Platform services, but I can't. I must set the `GOOGLE_CLOUD_PROJECT` env var or explicitly pass a project ID to a client for google-cloud-ruby to work.
## Steps to Reproduce
1. Install the Google Cloud SDK and Ruby
2. Run `gcloud init`, you'll be prompted to set a default project
3. Run `gcloud beta auth application-default login` and follow the OAuth 2.0 flow to login
4. Run `gem install google-cloud-storage` to install a google-cloud-ruby client library
5. Run `irb` and use the google-cloud-storage gem to interact with Google Cloud Storage.  In this case, I am following the example code snippet for [Retrieving Buckets](https://googlecloudplatform.github.io/google-cloud-ruby/#/docs/google-cloud/v0.20.0/google/cloud/storage) found on the website for google-cloud-ruby.

``` ruby
>> require "google/cloud"
=> true
>> gcloud = Google::Cloud.new
=> #<Object:0x007fc27b105f78 @project=nil, @keyfile=nil, @retries=nil, @timeout=nil>
>> storage = gcloud.storage
ArgumentError: project is missing
    from /Users/remily/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/google-cloud-storage-0.20.0/lib/google-cloud-storage.rb:110:in `storage'
    from /Users/remily/.rbenv/versions/2.3.1/lib/ruby/gems/2.3.0/gems/google-cloud-storage-0.20.0/lib/google-cloud-storage.rb:66:in `storage'
    from (irb):3
    from /Users/remily/.rbenv/versions/2.3.1/bin/irb:11:in `<main>'
```
## Expected

After a user runs `gcloud beta auth application-default login`, google-cloud-ruby client libraries should work without explicitly setting a project ID via the code or an environment variable.

**Comment by blowmage** [comment]
Thanks for the issue, @remi. I was not aware this was being discussed on the other language projects.

In GoogleCloudPlatform/google-cloud-node#1653 it seems like the consensus is that this functionality should be provided by [the authentication library](https://github.com/google/google-auth-library-nodejs/issues/97), and not the google-cloud-node library. For ruby, that is the googleauth gem. The googleauth gem does sniff for both the GCE environment, and the gcloud SDK Application Default Credentials. But when we asked them to support [retrieving the default project_id for GCE](https://github.com/google/google-auth-library-ruby/issues/43) they declined, saying that it was not something the library should provide.

What is your opinion? Is this something that googleuth should provide, as it looks to be in future releases for python and node? Or something that google-cloud should provide?

**Comment by remi** [comment]
Because this feature is so tightly coupled to the Google Cloud SDK command-line tool, I think it makes sense for this feature to be implemented in google-cloud rather than the generalized Ruby Google API client.

For Python, project ID detection is currently implemented in google-cloud-python, not pushed down into the general Python Google API client.

I rather like Python's project [detection](https://github.com/GoogleCloudPlatform/google-cloud-python/blob/037944198df6be151973d5fa7fa10852c1829153/core/google/cloud/_helpers.py#L286-L297):

 > 1. GOOGLE_CLOUD_PROJECT environment variable
 > 1. GOOGLE_APPLICATION_CREDENTIALS JSON file
 > 1. Get default service project from `$ gcloud beta auth application-default login`
 > 1. Google App Engine application ID (*not relevant for Ruby*)
 > 1. Google Compute Engine project ID (from metadata server)

Also, I see that googleauth [sniffs the Google Compute Engine metadata server](https://github.com/google/google-auth-library-ruby/blob/37ed189b2e5165243918fcde127ebe323e9a06b5/lib/googleauth/compute_engine.rb#L51-L56) for an [auth token](https://cloud.google.com/compute/docs/access/create-enable-service-accounts-for-instances#applications), which is very relevant to auth, I think that it makes the most sense for google-cloud-ruby to lookup the [project ID](https://cloud.google.com/compute/docs/storing-retrieving-metadata#default).

What do you think?

**Comment by remi** [comment]
**Follow-up comment:** I have been discussing this offline with Python and Node.js stakeholders.  The [referenced Node.js issue](https://github.com/GoogleCloudPlatform/google-cloud-node/issues/1653) also includes [discussion](https://github.com/GoogleCloudPlatform/google-cloud-node/issues/1653#issuecomment-250794512) about upcoming changes to the the Python Google Auth library to **move project detection into the authentication library**.

Reasoning / benefits:
 1. That Auth Library already has code that is specific to Google Cloud use-cases, such as Google Compute Engine use-cases and default credential lookup
 1. Having the code in one place is ideal for if/when any changes are made, eg. to config file format/location/etc
 1. Having this logic live outside of an API client library will allow any client-library to make use of this functionality, eg. the existing [Google API client for Ruby](https://github.com/google/google-api-ruby-client) will be able to make use of this feature as well

The benefits of putting this code into the [Google Auth library for Ruby](https://github.com/google/google-auth-library-ruby) are great and will benefit all libraries that make use of googleauth

I'll follow-up with an issue files on the googleauth library.  We can block this pending support for project detection in googleauth?

/cc @jonparrott @jmdobry

**Comment by blowmage** [comment]
Sounds good to me! We noticed that the newer keyfile JSON files also include the project_id. Since the google-cloud libs already provide googleauth with the JSON file I think it would be great if it gave back the project_id as well as the credentials object.

**Comment by blowmage** [comment]
This issue is blocked by google/google-auth-library-ruby#90, so I am moving this issue to the [feature backlog](https://github.com/GoogleCloudPlatform/google-cloud-ruby/wiki/Feature-backlog) until we are unblocked.

**Comment by AteqEjaz** [comment]
Still having the same issue with Ruby version 2.4.4 thanks

**Comment by blowmage** [comment]
The recently released [`googleauth` 0.7.0](https://rubygems.org/gems/googleauth/versions/0.7.0) will support the functionality to extract the `project_id` from service account credentials, but both `google-api-client` and `google-gax` pin the dependency on `googleauth` to a version lower than 0.7.0. We will have to coordinate the loosening of the downstream dependencies before we can start making use of this feature in the `google-cloud-*` gems.

**Comment by blowmage** [comment]
I'm going to reclose this issue and open a new issue for the remaining work.
